lib/puppet/metatype/manager.rb

```diff
@@ -143,29 +143,29 @@ def rmtype(name)
   #
   def type(name)
     # Avoid loading if name obviously is not a type name
-    if name.to_s.include?(':')
-      return nil
-    end
+    return nil if name.to_s.include? ':'
 
     @types ||= {}
 
     # We are overwhelmingly symbols here, which usually match, so it is worth
     # having this special-case to return quickly.  Like, 25K symbols vs. 300
     # strings in this method. --daniel 2012-07-17
-    return @types[name] if @types[name]
+    return @types[name] if @types.include? name
 
     # Try mangling the name, if it is a string.
     if name.is_a? String
       name = name.downcase.intern
-      return @types[name] if @types[name]
+      return @types[name] if @types.include? name
     end
     # Try loading the type.
     if typeloader.load(name, Puppet.lookup(:current_environment))
       Puppet.warning "Loaded puppet/type/#{name} but no class was created" unless @types.include? name
+    else
+      @types[name] = nil
```

Comment:
This is the only part of this commit that's a functional change, correct? And yes, I think agree with the comment in the PR description to the effect that we would need a "code changed" hook for this to be safe. I think.

Btw, the spec failure is fixed by adding `next unless klass` to the failing bit of `lib/puppet/reference/providers.rb`. Basically the current code won't ever return a `nil` to that site, but this change makes that possible.